### PR TITLE
Connector service proposal

### PR DESCRIPTION
**Pull request by lszymik**
Created on 2019-01-30 by Lukasz Szymik (@lszymik ).

## Decision log

| Name | Description |
|-----------------------|------------------------------------------------------------------------------------|
| Title | Central Connector Service|
| Ultimate decision maker(s) | Kyma Council |
| Due date | 2019-02-28 |
| Status | Proposed on 2019-01-30|
| Input provider(s) | me |
| Group(s) affected by the decision | all |
| Decision type | `Prioritization` |
| Earliest date to revisit the decision | 2020-02-28 |
| Affected decisions | `None` |

## Context

The proposal contains design and recommendation how we can evolve Connector service from Application Connector to work in the central mode overarching several Kyma clusters.

## Decision
Extend the Connector Service to be able to work as the central component?

## Consequences
- Possibility to provision certificates for multiple Kyma clusters using single infrastructure.

### CODE REVIEW

File: sig-and-wg/sig-core/proposals/two-phase-connector-service.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+1. An application calls info endpoint.
+2. The information about Kyma runtime is returned. The information contains: App Registry URL, Event service URL, Connector service in Runtime (with connection token.)
+3. An application is updating its configuration if something changed (e.g., new Event Service URL, the certificate is expiring, and a new one must be fetched.)
+4. The calls should be executed periodically (pulling). It will be used as a heartbeat from the connected application.
+
+**Security:** The client certificate is representing the connected application.
+
+#### Application is connecting to the Kyma runtime
+
+1. An application is requesting CSR information by sending registration token returned in the first phase.
```
Comment (pbochynski) [inline sig-and-wg/sig-core/proposals/two-phase-connector-service.md]: Do you mean a connection token mentioned in the previous section (point 2)?
How the information about connection/registration token is trusted?

File: sig-and-wg/sig-core/proposals/two-phase-connector-service.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+1. The Connector service handles client certificate provisioning for a Kyma runtime using the first client certificate.
+1. The new client certificate can be requested using the previous one (certificate rotation)
+1. The Connector service can trigger the creation of the application custom resource if it doesn't exist yet.
+
+## Suggested solution
+
+Connector service is extended to work with two phases:
+
+  - In the first phase, the Connector service is returning cluster metadata for CSR signing and information about App Registry URL.
+  - In the second phase, the Connector service is returning metadata of the Kyma runtime cluster for which a second client certificate can be obtained.
```
Comment (pbochynski) [inline sig-and-wg/sig-core/proposals/two-phase-connector-service.md]: I was thinking about it a little bit and I have doubts if the second client certificate will help in any aspect. In the beginning, we wanted to separate registration / metadata calls from actual workload (events). When I started to think about handling security issues (mainly client certificate is compromised), I came to the point that we have always invalidate both certificates. In this case, 2 certificates just bring unnecessary complexity.

File: sig-and-wg/sig-core/proposals/central-connector-service.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    kind: Secret
+    metadata:
+      name: nginx-auth-ca
+      namespace: kyma-integration
+    type: Opaque
+    EOF
+    kubectl -n kyma-system delete po -l "app=nginx-ingress"
+    ```
+
+6. Wait for the Pod to restart. Access the cluster using the previously generated `client.crt`:
```
Comment (tomekpapiernik) [inline sig-and-wg/sig-core/proposals/central-connector-service.md]: What Pod? Not really clear.

File: sig-and-wg/sig-core/proposals/central-connector-service.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+    ```
+    curl  https://gateway.{CLUSTER_DOMAIN}/test-application/v1/metadata/services --cert client.crt --key client.key
+    ```
+
+7. Revert changes by setting the root CA-signed certificate as the original one and call the cluster using of the intermediate certificate as the client certificate.
```
Comment (tomekpapiernik) [inline sig-and-wg/sig-core/proposals/central-connector-service.md]: I get the gist of it, but maybe it'd be better to show what commands to execute this action. Split into two steps to make it clearer if necessary.

File: sig-and-wg/sig-core/proposals/two-phase-connector-service.md (hunk 1 above)
Comment (lszymik) [reply sig-and-wg/sig-core/proposals/two-phase-connector-service.md]: Done

File: sig-and-wg/sig-core/proposals/central-connector-service.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+Proposed on 2018-12-17.
+
+## Motivation
+
+The Connector service is responsible for establishing a secure connection between the connected external solutions and Kyma runtime. Such a connection is achieved by providing the client certificate to the connected solution. This certificate is later validated by the Application Connector. The client certificate is used for registering application metadata, such as APIs, and sending events to Kyma.
+Currently, the connection between an external solution and Kyma is always point-to-point. 
+
+As the customers work with multiple Kyma clusters, they will benefit from extending the provisioning of Kyma client certificates. A central Connector Service would manage the provisioning of certificates for multiple Kyma clusters and connected clients. Such approach allows the users to control their entire Kyma ecosystem from a single, central point.
+Additionally, creating a central Connector Service enables the development of a central Application Registry. The central Application Registry can be used in a similar manner, serving multiple Kyma clusters and their connected external solutions.
```
Comment (pbochynski) [inline sig-and-wg/sig-core/proposals/central-connector-service.md]:
```suggestion
```

File: sig-and-wg/sig-core/proposals/central-connector-service.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+
+## Goal
+
+1. The Connector service handles client certificate provisioning for the connection with the Application Registry.
+2. The Connector service handles client certificate provisioning for the connection with the Event Service.
+3. The Connector service handles certificate provisioning for Kyma runtime.
+4. The Connector service handles certificate rotation.
+5. The Connector service returns information about the available cluster endpoints.
```
Comment (pbochynski) [inline sig-and-wg/sig-core/proposals/central-connector-service.md]: Please define Kyma runtime = cluster. Be more consistent.

**Pull request merged**